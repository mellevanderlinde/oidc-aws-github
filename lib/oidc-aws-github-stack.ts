import { Stack, StackProps, aws_iam as iam } from "aws-cdk-lib";
import { Construct } from "constructs";

export class OidcAwsGithubStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const oidcProvider = new iam.CfnOIDCProvider(this, "GithubOidcProvider", {
      url: "https://token.actions.githubusercontent.com",
      clientIdList: ["sts.amazonaws.com"],
    });

    const conditions: iam.Conditions = {
      StringEquals: {
        "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
        "token.actions.githubusercontent.com:sub":
          "repo:mellevanderlinde/oidc-aws-github:ref:refs/heads/main",
      },
    };

    const oidcRole = new iam.Role(this, "GithubOidcRole", {
      roleName: "github-oidc-role",
      assumedBy: new iam.WebIdentityPrincipal(oidcProvider.ref, conditions),
    });

    const lookupRole = iam.Role.fromRoleName(
      this,
      "LookupRole",
      `cdk-hnb659fds-lookup-role-${this.account}-${this.region}`,
    );

    lookupRole.grantAssumeRole(oidcRole);
  }
}
