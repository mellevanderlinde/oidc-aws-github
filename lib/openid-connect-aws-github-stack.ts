import { Stack, StackProps, aws_iam as iam } from "aws-cdk-lib";
import { Construct } from "constructs";

export class OpenidConnectAwsGithubStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const oidcProvider = new iam.CfnOIDCProvider(this, "GithubOidcProvider", {
      url: "https://token.actions.githubusercontent.com",
    });

    const conditions: iam.Conditions = {
      StringEquals: {
        "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
        "token.actions.githubusercontent.com:sub":
          "repo:mellevanderlinde/openid-connect-aws-github:ref:refs/heads/non-existing-branch", // TODO change to main branch, first verify that this fails
      },
    };

    new iam.Role(this, "GithubOidcRole", {
      roleName: "github-oidc-role",
      assumedBy: new iam.WebIdentityPrincipal(oidcProvider.ref, conditions),
    });
  }
}
