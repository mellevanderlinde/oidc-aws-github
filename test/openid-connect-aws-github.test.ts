import { App, assertions } from "aws-cdk-lib";
import { OpenidConnectAwsGithubStack } from "../lib/openid-connect-aws-github-stack";

test("Match with snapshot", () => {
  const app = new App();
  const stack = new OpenidConnectAwsGithubStack(
    app,
    "OpenidConnectAwsGithubStack",
  );
  const template = assertions.Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});
