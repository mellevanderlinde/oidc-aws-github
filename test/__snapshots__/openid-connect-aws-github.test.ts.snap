// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match with snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GithubOidcProvider": {
      "Properties": {
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "AWS::IAM::OIDCProvider",
    },
    "GithubOidcRole20798CD3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                  "token.actions.githubusercontent.com:sub": "repo:mellevanderlinde/openid-connect-aws-github:ref:refs/heads/non-existing-branch",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Ref": "GithubOidcProvider",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "github-oidc-role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
